specification.md

```diff
@@ -604,263 +514,687 @@ retries: []
+**(4) Event State Completes  Execution**: 
 
-The `keepActive` property allows you to change the default behavior of workflow instances.
-By default, as described in the [Core Concepts](#Core-Concepts) section, a workflow instance is terminated once there are no more 
-active execution paths, one of its active paths ends in a "terminate" [end definition](#End-Definition), or when
-its [`execTimeout`](#ExecTimeout-Definition) time is reached. 
+When our event state finishes its execution, the states "stateFilter" "output" filter expression is executed
+to filter the state data to create the final state data output.
 
-Setting the `keepActive` property to "true" allows you change this default behavior in that a workflow instance
-created from this workflow definition can only be terminated if one of its active paths ends in a "terminate" [end definition](#End-Definition), or when
-its [`execTimeout`](#ExecTimeout-Definition) time is reached. 
-This allows you to explicitly model workflows where an instance should be kept alive, to collect (event) data for example.
+Because our event state is also an end state, its data output becomes the final [workflow data output](#Workflow-data-output). Namely:
 
-You can reference the [specification examples](#Examples) to see the `keepActive` property in action.
+```json
+{
+   "finalCustomerGreeting": "Hola John Michaels!"
+}
+```
 
-#### ExecTimeout Definition
+#### Data Merging
 
-| Parameter | Description | Type | Required |
-| --- | --- | --- | --- |
-| interval | Timeout interval (ISO 8601 duration format) | string | yes |
-| interrupt | If `false`, workflow instance is allowed to finish current execution. If `true`, current workflow execution is stopped immediately. Default is `false`  | boolean | no |
-| runBefore | Name of a workflow state to be executed before workflow instance is terminated | string | no |
+Consumed event data (payload) as well as action execution results should be merged into the state data.
+Event and action data filters can be used to give more details about this operation.
 
+By default (no data filters specified), when an event is consumed, its entire data section (payload) should be merged
+to the state data. The merge should be applied to entire state data JSON element.
 
-<details><summary><strong>Click to view example definition</strong></summary>
-<p>
+In case of event and action filters, their "toStateData" property can be defined to select a specific element 
+of the state data with which the merge should be done against. If this element does not exist, a new one should
+be created first.
 
-<table>
-<tr>
-    <th>JSON</th>
-    <th>YAML</th>
-</tr>
-<tr>
-<td valign="top">
+When merging, the state data element and the data (payload)/action result should have the same type, meaning
+that you should not merge arrays with objects or objects with arrays etc. 
+
+When merging elements of type object should be done by inserting all the key-value pairs from both objects into
```

Comment:
the main driver to keep the workflow data as small as possible (for performance) is the state filters
i think a good practice for big event data/ large function results would be to first trim them down using data filters before merging them to state data

Reply:
This deep-merging behavior is complex, and leads to failure modes that users won't expect like being surprised by a runtime error when they merge 2 data-structures that overlap, but some of the data-types are different (array vs. object vs string etc).

I suspect most users would naturally expect a shallow merge and that very few users would actually want/need deep-merging.  Further, it adds new failure modes that the users have to watch for (like the disparate types problem) and adds a lot of complexity for runtimes to implement.

Finally, if a user really wants to deep-merge some data they've collected from multiple actions, if we were to only support shallow-merge then they can still do the deep-merge via fromStateData before passing the merged data to an action.

Reply:
ok that sounds reasonable

Reply:
This still seems to be describing a somewhat-deep merge.  I'm confused on whether we've resolved this or not.   

I'd like to propose the following (happy to move this to a follow-up issue if you prefer):

* Support simple insert, no merging.
* Add the ability for customers to reference the state data in their actionDataFilter.results or eventDataFilter.data expressions, which gives them the power to specify custom 'merge' behavior.

__WHY?__

Because this is much simpler to understand, there are no surprises about deep or shallow merging its just a simple insert by default.  It also lets users leverage the power of JQ to do custom 'merge' strategies.

Consider the example:
```yaml
results: ${.}
toStateData: .customer
```

This would simply insert the action result at the location .customer in state data.  If something already exists there in state data, this would completely over-write that data.

__What if users really want to do merge?__

We can give users much more flexibility, though it requires that the runtime make the state data available as a pre-defined jq variable.

With that we can give the user all kinds of flexibility.

(These examples assume the action result, referenced by `.` is a customer object)

Here's an example where we do a simple merge via actionDataFilter:
```yaml
results: ${$stateData.customer + .}
toStateData: .customer
```

The above example takes customer from stateData and shallow merges it with the action result, reassigning the results back to .customer in the state data.

Here's an example where we do a recursive merge via actionDataFilter:
```yaml
results: ${$stateData.customer * .}
toStateData: .customer
```

The above example takes customer from stateData and recursively merges it with the action result, reassigning the results back to .customer in the state data.

Reply:
I went ahead and made a separate issue for this: #286